docs/ErrorHandling.md

```diff
@@ -0,0 +1,745 @@
+    reasonable general assumption.
+-   `AVFoundation` provides a couple methods returning
+    `AVKeyValueStatus`. These produce an error if the API returned
+    `AVKeyValueStatusFailed`, which, interestingly enough, is not the
+    zero value.
+
+The clang attribute would specify how to test the return value for an
+error. For example:
+
+```objc
++ (NSInteger)writePropertyList:(id)plist
+                      toStream:(NSOutputStream *)stream
+                        format:(NSPropertyListFormat)format
+                       options:(NSPropertyListWriteOptions)opt
+                         error:(out NSError **)error
+    NS_ERROR_RESULT(0);
+
+- (AVKeyValueStatus)statusOfValueForKey:(NSString *)key
+                                  error:(NSError **)
+    NS_ERROR_RESULT(AVKeyValueStatusFailed);
+```
+
+We should also provide a Clang attribute which specifies that the
+correct way to test for an error is to check the out-parameter. Both of
+these attributes could potentially be used by the static analyzer, not
+just Swift. (For example, they could try to detect an invalid error
+check.)
+
+Cases that do not match the automatically imported patterns and that
+lack an attribute would be left unmodified (i.e., they\'d keep their
+NSErrorPointer argument) and considered \"not awesome\" in the SDK
+auditing tool. These will still be usable in Swift: callers will get the
+NSError back like they do today, and have to throw the result manually.
+
+For initializers, importing an initializer as throwing takes precedence
+over importing it as failable. That is, an imported initializer with a
+nullable result and an error parameter would be imported as throwing.
+Throwing initializers have very similar constraints to failable
+initializers; in a way, it\'s just a new axis of failability.
+
+One limitation of this approach is that we need to be able to
+reconstruct the selector to use when an overload of a method is
+introduced. For this reason, the import is likely to be limited to
+methods where the error parameter is the last one and the corresponding
+selector chunk is either `error:` or the first chunk (see below).
+Empirically, this seems to do the right thing for all but two sets of
+APIs in the public API:
+
+-   The `ISyncSessionDriverDelegate` category on `NSObject` declares
+    half-a-dozen methods like this:
+
+    ```objc
+    - (BOOL)sessionDriver:(ISyncSessionDriver *)sender
+            didRegisterClientAndReturnError:(NSError **)outError;
+    ```
+
+    Fortunately, these delegate methods were all deprecated in Lion, and
+    are thus unavailable in Swift.
+
+-   `NSFileCoordinator` has half a dozen methods where the `error:`
+    clause is second-to-last, followed by a block argument. These
+    methods are not deprecated as far as I know.
+
+The above translation rule would import methods like this one from
+`NSDocument`:
+
+    ```objc
+    - (NSDocument *)duplicateAndReturnError:(NSError **)outError;
+    ```
```

Comment:
This and the next three code blocks are not rendered properly, probably due to the indentation.